File: PasswordStorage.cs

using System;
using System.Text;
using System.Security.Cryptography;

namespace PasswordSecurity
{
    class InvalidHashException : Exception
    {
        public InvalidHashException() { }
        public InvalidHashException(string message)
            : base(message) { }
        public InvalidHashException(string message, Exception inner)
            : base(message, inner) { }
    }

    class CannotPerformOperationException : Exception
    {
        public CannotPerformOperationException() { }
        public CannotPerformOperationException(string message)
            : base(message) { }
        public CannotPerformOperationException(string message, Exception inner)
            : base(message, inner) { }
    }

    class PasswordStorage
    {
        // These constants may be changed without breaking existing hashes.
        public const int SALT_BYTES = 24;
        public const int HASH_BYTES = 18;
        public const int PBKDF2_ITERATIONS = 64000;

        // These constants define the encoding and may not be changed.
        public const int HASH_SECTIONS = 5;
        public const int HASH_ALGORITHM_INDEX = 0;
        public const int ITERATION_INDEX = 1;
        public const int HASH_SIZE_INDEX = 2;
        public const int SALT_INDEX = 3;
        public const int PBKDF2_INDEX = 4;

        public static string CreateHash(string password)
        {
            // Generate a random salt
            byte[] salt = new byte[SALT_BYTES];
            try {
                using (RNGCryptoServiceProvider csprng = new RNGCryptoServiceProvider()) {
                        csprng.GetBytes(salt);
                }
            } catch (CryptographicException ex) {
                throw new CannotPerformOperationException(
                    "Random number generator not available.",
                    ex
                );
            } catch (ArgumentNullException ex) {
                throw new CannotPerformOperationException(
                    "Invalid argument given to random number generator.",
                    ex
                );
            }

            byte[] hash = PBKDF2(password, salt, PBKDF2_ITERATIONS, HASH_BYTES);

            // format: algorithm:iterations:hashSize:salt:hash
            String parts = "sha1:" +
                PBKDF2_ITERATIONS +
                ":" +
                hash.Length +
                ":" +
                Convert.ToBase64String(salt) +
                ":" +
                Convert.ToBase64String(hash);
            return parts;
        }

        public static bool VerifyPassword(string password, string goodHash)
        {
            char[] delimiter = { ':' };
            string[] split = goodHash.Split(delimiter);

            if (split.Length != HASH_SECTIONS) {
                throw new InvalidHashException(
                    "Fields are missing from the password hash."
                );
            }

            // We only support SHA1 with C#.
            if (split[HASH_ALGORITHM_INDEX] != "sha1") {
                throw new CannotPerformOperationException(
                    "Unsupported hash type."
                );
            }

            int iterations = 0;
            try {
                iterations = Int32.Parse(split[ITERATION_INDEX]);
            } catch (ArgumentNullException ex) {
                throw new CannotPerformOperationException(
                    "Invalid argument given to Int32.Parse",
                    ex
                );
            } catch (FormatException ex) {
                throw new InvalidHashException(
                    "Could not parse the iteration count as an integer.",
                    ex
                );
            } catch (OverflowException ex) {
                throw new InvalidHashException(
                    "The iteration count is too large to be represented.",
                    ex
                );
            }

            if (iterations < 1) {
                throw new InvalidHashException(
                    "Invalid number of iterations. Must be >= 1."
                );
            }

            byte[] salt = null;
            try {
                salt = Convert.FromBase64String(split[SALT_INDEX]);
            }
            catch (ArgumentNullException ex) {
                throw new CannotPerformOperationException(
                    "Invalid argument given to Convert.FromBase64String",
                    ex
                );
            } catch (FormatException ex) {
                throw new InvalidHashException(
                    "Base64 decoding of salt failed.",
                    ex
                );
            }

            byte[] hash = null;
            try {
                hash = Convert.FromBase64String(split[PBKDF2_INDEX]);
            }
            catch (ArgumentNullException ex) {
                throw new CannotPerformOperationException(
                    "Invalid argument given to Convert.FromBase64String",
                    ex
                );
            } catch (FormatException ex) {
                throw new InvalidHashException(
                    "Base64 decoding of pbkdf2 output failed.",
                    ex
                );
            }

            int storedHashSize = 0;
            try {
                storedHashSize = Int32.Parse(split[HASH_SIZE_INDEX]);
            } catch (ArgumentNullException ex) {
                throw new CannotPerformOperationException(
                    "Invalid argument given to Int32.Parse",
                    ex
                );
            } catch (FormatException ex) {
                throw new InvalidHashException(
                    "Could not parse the hash size as an integer.",
                    ex
                );
            } catch (OverflowException ex) {
                throw new InvalidHashException(
                    "The hash size is too large to be represented.",
                    ex
                );
            }

            if (storedHashSize != hash.Length) {
                throw new InvalidHashException(
                    "Hash length doesn't match stored hash length."
                );
            }

            byte[] testHash = PBKDF2(password, salt, iterations, hash.Length);
            return SlowEquals(hash, testHash);
        }

        private static bool SlowEquals(byte[] a, byte[] b)
        {
            uint diff = (uint)a.Length ^ (uint)b.Length;
            for (int i = 0; i < a.Length && i < b.Length; i++) {
                diff |= (uint)(a[i] ^ b[i]);
            }
            return diff == 0;
        }

        private static byte[] PBKDF2(string password, byte[] salt, int iterations, int outputBytes)
        {
            using (Rfc2898DeriveBytes pbkdf2 = new Rfc2898DeriveBytes(password, salt)) {
                pbkdf2.IterationCount = iterations;
                return pbkdf2.GetBytes(outputBytes);
            }
        }
    }
}

File: PasswordStorage.vb
Imports System
Imports System.Text
Imports System.Security.Cryptography

Namespace PasswordSecurity

    Class InvalidHashException
        Inherits Exception

        ' Default constructor
        Sub New()
        End Sub

        Sub New(message As String)
            MyBase.New(message)
        End Sub

        Sub New(message As String, innerException As Exception)
            MyBase.New(message, innerException)
        End Sub

    End Class

    Class CannotPerformOperationException
        Inherits Exception

        ' Default constructor
        Sub New()
        End Sub

        Sub New(message As String)
            MyBase.New(message)
        End Sub

        Sub New(message As String, innerException As Exception)
            MyBase.New(message, innerException)
        End Sub
    End Class

    Public Class PasswordStorage

        'These constants may be changed without breaking existing hashes.
        Public Const SALT_BYTES As Integer = 24
        Public Const HASH_BYTES As Integer = 18
        Public Const PBKDF2_ITERATIONS As Integer = 64000

        ' These constants define the encoding And may Not be changed.
        Public Const HASH_SECTIONS As Integer = 5
        Public Const HASH_ALGORITHM_INDEX As Integer = 0
        Public Const ITERATION_INDEX As Integer = 1
        Public Const HASH_SIZE_INDEX As Integer = 2
        Public Const SALT_INDEX As Integer = 3
        Public Const PBKDF2_INDEX As Integer = 4

        Public Shared Function CreateHash(password As String) As String

            Dim salt(SALT_BYTES) As Byte
            Try

                Using csprng = New RNGCryptoServiceProvider()
                    csprng.GetBytes(salt)
                End Using

            Catch ex As CryptographicException
                Throw New CannotPerformOperationException("Random number generator not available", ex)
            Catch ex As ArgumentException
                Throw New CannotPerformOperationException("Invalid argument given to random number generator", ex)
            End Try

            Dim hash() As Byte = PBKDF2(password, salt, PBKDF2_ITERATIONS, HASH_BYTES)

            ' format: algorithm:iteration:hashSize:salt:hash
            Dim parts As String = "sha1:" + PBKDF2_ITERATIONS.ToString() + ":" + hash.Length.ToString() + ":" + Convert.ToBase64String(salt) + ":" + Convert.ToBase64String(hash)
            Return parts

        End Function

        Public Shared Function VerifyPassword(password As String, goodHash As String) As Boolean
            Dim delimter() As Char = {":"}
            Dim split As String() = goodHash.Split(delimter)

            If split.Length <> HASH_SECTIONS Then
                Throw New InvalidHashException("Fields are missing from the password hash")
            End If

            ' We only support SHA1 with VB.NET
            If split(HASH_ALGORITHM_INDEX) <> "sha1" Then
                Throw New CannotPerformOperationException("Unsupported hash type.")
            End If

            Dim iterations As Integer = 0
            Try
                iterations = Integer.Parse(split(ITERATION_INDEX))
            Catch ex As ArgumentException
                Throw New CannotPerformOperationException("Invalid argument given to Int32.Parse", ex)
            Catch ex As FormatException
                Throw New InvalidHashException("Could not parse the iteration count as an integer.", ex)
            Catch ex As OverflowException
                Throw New InvalidHashException("The iteration count is too large to be represented.", ex)
            End Try

            If iterations < 1 Then
                Throw New InvalidHashException("Invalid number of interations. Must be >= 1.")
            End If

            Dim salt() As Byte = Nothing
            Try
                salt = Convert.FromBase64String(split(SALT_INDEX))
            Catch ex As ArgumentException
                Throw New CannotPerformOperationException("Invalid argument given to Convert.FromBase64String.", ex)
            Catch ex As FormatException
                Throw New InvalidHashException("Base64 decoding of salt failed.", ex)
            End Try

            Dim hash() As Byte = Nothing
            Try
                hash = Convert.FromBase64String(split(PBKDF2_INDEX))
            Catch ex As ArgumentException
                Throw New CannotPerformOperationException("Invalid argument given to Convert.FromBase64String.", ex)
            Catch ex As FormatException
                Throw New InvalidHashException("Base64 decoding of pbkdf2 output failed.", ex)
            End Try

            Dim storedHashSize As Integer = 0
            Try
                storedHashSize = Integer.Parse(split(HASH_SIZE_INDEX))
            Catch ex As ArgumentException
                Throw New CannotPerformOperationException("Invalid argument given to Integer.Parse.", ex)
            Catch ex As FormatException
                Throw New InvalidHashException("The hash size is too large to be represented.", ex)
            End Try

            If storedHashSize <> hash.Length Then
                Throw New InvalidHashException("Hash length doesn't match stored hash length.")
            End If

            Dim testHash() As Byte = PBKDF2(password, salt, iterations, hash.Length)
            Return SlowEquals(hash, testHash)
        End Function

        Private Shared Function SlowEquals(a() As Byte, b() As Byte) As Boolean
            Dim diff As UInteger = CUInt(a.Length) Xor CUInt(b.Length)
            Dim i = 0
            While i < a.Length And i < b.Length
                diff = diff Or CUInt(a(i) Xor b(i))
                i = i + 1
            End While
            Return diff = 0
        End Function

        Private Shared Function PBKDF2(password As String, salt() As Byte, iterations As Integer, outputBytes As Integer)
            Using pbkdf2DerivedBytes = New Rfc2898DeriveBytes(password, salt)
                pbkdf2DerivedBytes.IterationCount = iterations
                Return pbkdf2DerivedBytes.GetBytes(outputBytes)
            End Using
        End Function

    End Class

End Namespace


